astropy/utils/metadata.py

```diff
@@ -414,3 +415,73 @@ def __set__(self, instance, value):
                     instance._meta = value
             else:
                 raise TypeError("meta attribute must be dict-like")
+
+
+class MetaAttribute:
+    """
+    Descriptor to define custom attribute which gets stored in the object
+    ``meta`` dict and can have a defined default.
+
+    This descriptor is intended to provide a convenient way to add attributes
+    to a subclass of a complex class such as ``Table`` or ``NDData``.
+
+    This requires that the object has an attribute ``meta`` which is a
+    dict-like object.  The value of the MetaAttribute will be stored in a
+    new dict meta['__attributes__'] that is created when required.
+
+    Classes that define MetaAttributes are encouraged to support initializing
+    the attributes via the class ``__init__``.  For example::
+
+        for attr in list(kwargs):
+            descr = getattr(self.__class__, attr, None)
+            if isinstance(descr, MetaAttribute):
+                setattr(self, attr, kwargs.pop(attr))
+
+    The name of a ``MetaAttribute`` cannot be the same as any of the following:
+
+    - Keyword argument in the owner class ``__init__``
+    - Method or attribute of the "parent class", where the parent class is
+      taken to be ``owner.__mro__[1]``.
+
+    :param default: default value
+
+    """
+    def __init__(self, default=None):
+        self.default = default
+
+    def __get__(self, instance, owner):
+        # When called without an instance, return self to allow access
+        # to descriptor attributes.
+        if instance is None:
+            return self
+
+        # Get the __attributes__ dict and create if not there already.
+        meta = instance.meta.setdefault('__attributes__', {})
+        try:
+            value = meta[self.name]
+        except KeyError:
+            if self.default is not None:
+                meta[self.name] = deepcopy(self.default)
+            # Return either specified default or None
+            value = meta.get(self.name)
+        return value
+
+    def __set__(self, instance, value):
+        # Get the __attributes__ dict and create if not there already.
+        meta = instance.meta.setdefault('__attributes__', {})
+        meta[self.name] = value
+
```

Comment:
Maybe define a `__del__` method while we are at it?

Reply:
I started down that path but realized that for this current implementation `__delete__` doesn't quite make sense.  You cannot actually delete a `MetaAttribute` because it always comes back with the default.  So you would get:
```
>>> del obj.meta_attr
>>> obj.meta_attr
# prints the default for meta_attr
```
Now we *could* change `__init__` so that there is a difference between a meta attribute that has an explicitly specified default and one that does not:
```
class MyClass:
    ma1 = MetaAttribute()
    ma2 = MetaAttribute(default=5)

obj = MyClass()
obj.ma2  # returns 5
obj.ma1  # AttributeError (current code returns None)
```
I'm not sure which way to go here.  In my real-world applications where this was developed, it was convenient to have that default of `None`.  OTOH it is a bit more Pythonic that accessing an attribute that has not been set should raise AttributeError unless a default is explicitly provided.  If we do this then `__delete__` would make sense.

Reply:
I started writing this implementation to see how it looks.  But I got really stuck at how confusing it seems in practice to have explicitly defined `ma1` as a class attribute and then immediately afterword get an `AttributeError: 'MyClass' has no attribute 'ma1'` when it is right there as a class attribute.

I'm kind of leaning back to these always having an implicit default of None, and thinking of them as a bit more like properties than attributes.  Maybe they should be called `MetaProperty`?

Reply:
And in fact when I started this whole thing for my own work, I actually wrote the first couple as properties and then realized that I could generalize and reduce boilerplate code by making the descriptor.

Reply:
For the record, this is what I started doing.
```
--- a/astropy/utils/metadata.py
+++ b/astropy/utils/metadata.py
@@ -446,8 +446,12 @@ class MetaAttribute:
     :param default: default value
 
     """
-    def __init__(self, default=None):
-        self.default = default
+    def __init__(self, **kwargs):
+        if 'default' in kwargs:
+            self.default = kwargs.pop('default')
+        if kwargs:
+            raise TypeError(f"__init__() got unexpected keyword argument "
+                            f"'{list(kwargs)[0]}'")
 
     def __get__(self, instance, owner):
         # When called without an instance, return self to allow access
@@ -460,10 +464,11 @@ class MetaAttribute:
         try:
             value = attributes[self.name]
         except KeyError:
-            if self.default is not None:
-                attributes[self.name] = deepcopy(self.default)
-            # Return either specified default or None
-            value = attributes.get(self.name)
+            if hasattr(self, 'default'):
+                value = attributes[self.name] = deepcopy(self.default)
+            else:
+                raise AttributeError(f"'{instance.__class__.__name__}' object has no "
+                                     f"attribute 'self.name'")
         return value
 
     def __set__(self, instance, value):
```